src/eth/TransactionManager.js

```diff
@@ -2,19 +2,57 @@ import PublicService from '../core/PublicService';
 import TransactionObject from './TransactionObject';
 import { Contract } from 'ethers';
 import { dappHub } from '../../contracts/abis';
+import { uniqueId } from '../utils';
+import { has } from 'lodash';
+import debug from 'debug';
+const log = debug('dai:testing');
 
 let txId = 1;
 
+// decorator definition
+export function tracksTransactions(target, name, descriptor) {
+  const original = descriptor.value;
+  descriptor.value = function(...args) {
+    const lastArg = args[args.length - 1];
+    let options;
+    if (typeof lastArg === 'object' && lastArg.constructor === Object) {
+      args = args.slice(0, args.length - 1);
+      options = lastArg;
+    } else {
+      options = {};
+    }
+
+    const promise = (async () => {
+      // this "no-op await" is necessary for the inner reference to the
+      // outer promise to become valid
+      await 0;
+
+      options.promise = promise;
```

Comment:
now we have a reference to the promise that we're inside! we can now pass it as an argument to the function we're wrapping, so that contract calls can pass it along to TransactionManager. i know, this is some Klein-bottle voodoo.